universal-login-relayer/lib/services/CounterfactualIdentityService.js

```diff
@@ -0,0 +1,125 @@
+
+    const counterfactualTransaction = this._counterfactualizeSignedTransaction(signedTransaction);
+
+    const fullCounterfactualData = await this._extractCounterfactualParams(counterfactualTransaction);
+
+    this.hooks.emit('counterfactuallyCreated', {address: fullCounterfactualData.counterfactualContractAddress});
+
+    return fullCounterfactualData;
+  }
+
+  _createDeployTransaction(managementKey, ensName) {
+    const key = addressToBytes32(managementKey);
+    const bytecode = `0x${Identity.bytecode}`;
+    const ensArgs = this.ensService.argsFor(ensName);
+    const args = [key, ...ensArgs];
+    return ethers.Contract.getDeployTransaction(bytecode, this.abi, ...args);
+  }
+
+  _setupDeployTransaction(deployTransaction, overrideOptions) {
+    const newDeployTransaction = {
+      ...deployTransaction,
+      ...defaultDeployOptions,
+      ...overrideOptions
+    };
+    return newDeployTransaction;
+  }
+
+  _signDeployTransaction(deployTransaction) {
+    return this.wallet.sign(deployTransaction);
+  }
+
+  _counterfactualizeSignedTransaction(signedTransaction) {
+    const signedTransNoRSV = signedTransaction.substring(0, signedTransaction.length - 134);
+
+    let randomS = utils.keccak256(utils.randomBytes(3));
+    randomS = `0 + ${randomS.substring(3, randomS.length)}`;
+
+    const counterfactualMagic = `1ba079be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798a0`;
+
+    return `${signedTransNoRSV}${counterfactualMagic}${randomS}`;
+  }
+
+  async _extractCounterfactualParams(counterfactualTransaction) {
+    const parsedTrans = Wallet.parseTransaction(counterfactualTransaction);
+
+    const counterfactualDeploymentPayer = parsedTrans.from;
+
+    const nonce = await this.provider.getTransactionCount(counterfactualDeploymentPayer);
+
+    const transaction = {
+      from: counterfactualDeploymentPayer,
+      nonce
+    };
+
+    const counterfactualContractAddress = utils.getContractAddress(transaction);
+
+    const counterfactualBalance = await this.provider.getBalance(counterfactualContractAddress);
+    if (utils.bigNumberify(counterfactualBalance).gt(0)) {
+      throw new Error('Contract with such address has already been used');
+    }
+
+    return {
+      counterfactualContractAddress,
+      counterfactualTransaction,
+      counterfactualDeploymentPayer
+    };
+  }
+
+  async deployProxy(fullCounterfactualData) {
+    const counterfactualBalance = await this.provider.getBalance(fullCounterfactualData.counterfactualContractAddress);
+
+    if (utils.bigNumberify(counterfactualBalance).lt(this.minimumWeiInContractBeforeDeploy)) {
+      throw new Error('Counterfactual contract should have ethers');
+    }
+
+    const from = fullCounterfactualData.counterfactualDeploymentPayer;
+
+    const existingValue = await this.provider.getBalance(from);
+
+    const valueToBeSent = this.deploymentValue.sub(existingValue);
+
+    const fundTransaction = await this.wallet.send(from, valueToBeSent);
```

Comment:
This requires two transactions, can we make it one with the dynamic factory?

Reply:
Not sure what do you mean by the dynamic factory, but this always requires two transactions:
1. Funding the address that is going to deploy
2. Deploy